File: javascript/src/metrics.ts
/**
 * Metrics collection for observability.
 *
 * Tracks request latency, error rates, queue depth, and circuit breaker events.
 */

/** Metrics for a single completed request. */
export interface RequestMetrics {
    requestId: string;
    func: string;
    namespace: string;
    success: boolean;
    latencyMs: number;
    error?: string;
    timestamp: number;
}

/** Point-in-time snapshot of all metrics. */
export interface MetricsSnapshot {
    // Counters
    requestsTotal: number;
    requestsSuccess: number;
    requestsFailed: number;

    // Latency (milliseconds)
    latencyAvgMs: number;
    latencyP50Ms: number;
    latencyP95Ms: number;
    latencyP99Ms: number;
    latencyMinMs: number;
    latencyMaxMs: number;

    // Queue
    queueDepth: number;
    queueMaxDepth: number;

    // Circuit breaker
    circuitBreakerTrips: number;
    circuitBreakerState: "closed" | "open" | "half-open";

    // Heartbeat
    heartbeatRttAvgMs: number;
    heartbeatRttLastMs: number;
    heartbeatMisses: number;

    // Time window
    windowSeconds: number;

    // Timestamp
    timestamp: number;
}

/** Metrics dictionary for JSON serialization. */
export interface MetricsDict {
    requests: {
        total: number;
        success: number;
        failed: number;
        errorRate: number;
    };
    latencyMs: {
        avg: number;
        p50: number;
        p95: number;
        p99: number;
        min: number;
        max: number;
    };
    queue: {
        depth: number;
        maxDepth: number;
    };
    circuitBreaker: {
        trips: number;
        state: string;
    };
    heartbeat: {
        rttAvgMs: number;
        rttLastMs: number;
        misses: number;
    };
}

/**
 * Thread-safe metrics collector for ParentWorker.
 *
 * Usage:
 *     const metrics = new Metrics();
 *
 *     // Start tracking a request
 *     const start = metrics.startRequest("req-123", "myFunc", "default");
 *     // ... do work ...
 *     metrics.endRequest(start, "req-123", true);
 *
 *     // Get snapshot
 *     const snapshot = metrics.snapshot();
 *     console.log(`Avg latency: ${snapshot.latencyAvgMs}ms`);
 */
export class Metrics {
    private readonly maxLatencySamples: number;
    private readonly windowSeconds: number;

    // Counters
    private _requestsTotal: number = 0;
    private _requestsSuccess: number = 0;
    private _requestsFailed: number = 0;
    private _circuitBreakerTrips: number = 0;
    private _circuitBreakerState: "closed" | "open" | "half-open" = "closed";

    // Queue tracking
    private _queueDepth: number = 0;
    private _queueMaxDepth: number = 0;

    // Latency samples (circular buffer via array slice)
    private _latencies: number[] = [];

    // Heartbeat RTT samples
    private _heartbeatRtts: number[] = [];
    private _heartbeatMisses: number = 0;

    constructor(options?: { maxLatencySamples?: number; windowSeconds?: number }) {
        this.maxLatencySamples = options?.maxLatencySamples ?? 1000;
        this.windowSeconds = options?.windowSeconds ?? 60.0;
    }

    /**
     * Start tracking a request.
     * Returns start timestamp for later endRequest() call.
     */
    startRequest(requestId: string, func: string, namespace: string = "default"): number {
        this._requestsTotal++;
        this._queueDepth++;
        this._queueMaxDepth = Math.max(this._queueMaxDepth, this._queueDepth);
        return performance.now();
    }

    /**
     * End tracking a request.
     * Returns latency in milliseconds.
     */
    endRequest(
        startTime: number,
        requestId: string,
        success: boolean = true,
        error?: string
    ): number {
        const latencyMs = performance.now() - startTime;

        this._queueDepth--;

        if (success) {
            this._requestsSuccess++;
        } else {
            this._requestsFailed++;
        }

        // Store latency sample (circular buffer behavior)
        this._latencies.push(latencyMs);
        if (this._latencies.length > this.maxLatencySamples) {
            this._latencies.shift();
        }

        return latencyMs;
    }

    /** Record a circuit breaker trip event. */
    recordCircuitBreakerTrip(): void {
        this._circuitBreakerTrips++;
        this._circuitBreakerState = "open";
    }

    /** Record circuit breaker reset (closed). */
    recordCircuitBreakerReset(): void {
        this._circuitBreakerState = "closed";
    }

    /** Record circuit breaker entering half-open state. */
    recordCircuitBreakerHalfOpen(): void {
        this._circuitBreakerState = "half-open";
    }

    /** Record a heartbeat round-trip time. */
    recordHeartbeatRtt(rttMs: number): void {
        this._heartbeatRtts.push(rttMs);
        if (this._heartbeatRtts.length > 100) {
            this._heartbeatRtts.shift();
        }
    }

    /** Record a missed heartbeat (timeout). */
    recordHeartbeatMiss(): void {
        this._heartbeatMisses++;
    }

    /** Get a point-in-time snapshot of all metrics. */
    snapshot(): MetricsSnapshot {
        const latencies = [...this._latencies];

        // Calculate percentiles
        let latencyAvg = 0;
        let latencyP50 = 0;
        let latencyP95 = 0;
        let latencyP99 = 0;
        let latencyMin = 0;
        let latencyMax = 0;

        if (latencies.length > 0) {
            const sorted = [...latencies].sort((a, b) => a - b);
            const n = sorted.length;
            const p50Idx = Math.floor(n * 0.5);
            const p95Idx = Math.floor(n * 0.95);
            const p99Idx = Math.floor(n * 0.99);

            latencyAvg = latencies.reduce((a, b) => a + b, 0) / n;
            latencyP50 = sorted[Math.min(p50Idx, n - 1)];
            latencyP95 = sorted[Math.min(p95Idx, n - 1)];
            latencyP99 = sorted[Math.min(p99Idx, n - 1)];
            latencyMin = sorted[0];
            latencyMax = sorted[n - 1];
        }

        // Calculate heartbeat RTT average
        let heartbeatRttAvg = 0;
        let heartbeatRttLast = 0;
        if (this._heartbeatRtts.length > 0) {
            heartbeatRttAvg = this._heartbeatRtts.reduce((a, b) => a + b, 0) / this._heartbeatRtts.length;
            heartbeatRttLast = this._heartbeatRtts[this._heartbeatRtts.length - 1];
        }

        return {
            requestsTotal: this._requestsTotal,
            requestsSuccess: this._requestsSuccess,
            requestsFailed: this._requestsFailed,
            latencyAvgMs: latencyAvg,
            latencyP50Ms: latencyP50,
            latencyP95Ms: latencyP95,
            latencyP99Ms: latencyP99,
            latencyMinMs: latencyMin,
            latencyMaxMs: latencyMax,
            queueDepth: this._queueDepth,
            queueMaxDepth: this._queueMaxDepth,
            circuitBreakerTrips: this._circuitBreakerTrips,
            circuitBreakerState: this._circuitBreakerState,
            heartbeatRttAvgMs: heartbeatRttAvg,
            heartbeatRttLastMs: heartbeatRttLast,
            heartbeatMisses: this._heartbeatMisses,
            windowSeconds: this.windowSeconds,
            timestamp: Date.now() / 1000,
        };
    }

    /** Reset all metrics. */
    reset(): void {
        this._requestsTotal = 0;
        this._requestsSuccess = 0;
        this._requestsFailed = 0;
        this._circuitBreakerTrips = 0;
        this._circuitBreakerState = "closed";
        this._queueDepth = 0;
        this._queueMaxDepth = 0;
        this._latencies = [];
        this._heartbeatRtts = [];
        this._heartbeatMisses = 0;
    }

    /** Get metrics as a dictionary (for logging/serialization). */
    toDict(): MetricsDict {
        const snapshot = this.snapshot();
        return {
            requests: {
                total: snapshot.requestsTotal,
                success: snapshot.requestsSuccess,
                failed: snapshot.requestsFailed,
                errorRate: snapshot.requestsTotal > 0
                    ? snapshot.requestsFailed / snapshot.requestsTotal
                    : 0.0,
            },
            latencyMs: {
                avg: Math.round(snapshot.latencyAvgMs * 100) / 100,
                p50: Math.round(snapshot.latencyP50Ms * 100) / 100,
                p95: Math.round(snapshot.latencyP95Ms * 100) / 100,
                p99: Math.round(snapshot.latencyP99Ms * 100) / 100,
                min: Math.round(snapshot.latencyMinMs * 100) / 100,
                max: Math.round(snapshot.latencyMaxMs * 100) / 100,
            },
            queue: {
                depth: snapshot.queueDepth,
                maxDepth: snapshot.queueMaxDepth,
            },
            circuitBreaker: {
                trips: snapshot.circuitBreakerTrips,
                state: snapshot.circuitBreakerState,
            },
            heartbeat: {
                rttAvgMs: Math.round(snapshot.heartbeatRttAvgMs * 100) / 100,
                rttLastMs: Math.round(snapshot.heartbeatRttLastMs * 100) / 100,
                misses: snapshot.heartbeatMisses,
            },
        };
    }
}

File: rust/src/metrics.rs
//! Metrics collection for observability.
//!
//! Tracks request latency, error rates, queue depth, and circuit breaker events.

use serde::{Deserialize, Serialize};
use std::collections::VecDeque;
use std::sync::Arc;
use std::time::Instant;
use tokio::sync::RwLock;

/// Metrics for a single completed request.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct RequestMetrics {
    pub request_id: String,
    pub function: String,
    pub namespace: String,
    pub success: bool,
    pub latency_ms: f64,
    pub error: Option<String>,
    pub timestamp: f64,
}

/// Point-in-time snapshot of all metrics.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct MetricsSnapshot {
    // Counters
    pub requests_total: usize,
    pub requests_success: usize,
    pub requests_failed: usize,

    // Latency (milliseconds)
    pub latency_avg_ms: f64,
    pub latency_p50_ms: f64,
    pub latency_p95_ms: f64,
    pub latency_p99_ms: f64,
    pub latency_min_ms: f64,
    pub latency_max_ms: f64,

    // Queue
    pub queue_depth: usize,
    pub queue_max_depth: usize,

    // Circuit breaker
    pub circuit_breaker_trips: usize,
    pub circuit_breaker_state: String,

    // Heartbeat
    pub heartbeat_rtt_avg_ms: f64,
    pub heartbeat_rtt_last_ms: f64,
    pub heartbeat_misses: usize,

    // Time window
    pub window_seconds: f64,

    // Timestamp
    pub timestamp: f64,
}

/// Thread-safe metrics collector for ParentWorker.
///
/// # Example
///
/// ```rust,no_run
/// use multifrost::metrics::Metrics;
///
/// let metrics = Metrics::new();
///
/// // Start tracking a request
/// let start = metrics.start_request("req-123", "myFunc", "default");
/// // ... do work ...
/// metrics.end_request(start, "req-123", true);
///
/// // Get snapshot
/// let snapshot = metrics.snapshot().await;
/// println!("Avg latency: {}ms", snapshot.latency_avg_ms);
/// ```
#[derive(Clone)]
pub struct Metrics {
    inner: Arc<RwLock<MetricsInner>>,
}

struct MetricsInner {
    max_latency_samples: usize,
    window_seconds: f64,

    // Counters
    requests_total: usize,
    requests_success: usize,
    requests_failed: usize,
    circuit_breaker_trips: usize,
    circuit_breaker_state: String,

    // Queue tracking
    queue_depth: usize,
    queue_max_depth: usize,

    // Latency samples (circular buffer)
    latencies: VecDeque<f64>,

    // Heartbeat RTT samples
    heartbeat_rtts: VecDeque<f64>,
    heartbeat_misses: usize,
}

impl Metrics {
    /// Create a new Metrics collector.
    pub fn new() -> Self {
        Self::with_options(1000, 60.0)
    }

    /// Create a new Metrics collector with custom options.
    pub fn with_options(max_latency_samples: usize, window_seconds: f64) -> Self {
        Self {
            inner: Arc::new(RwLock::new(MetricsInner {
                max_latency_samples,
                window_seconds,
                requests_total: 0,
                requests_success: 0,
                requests_failed: 0,
                circuit_breaker_trips: 0,
                circuit_breaker_state: "closed".to_string(),
                queue_depth: 0,
                queue_max_depth: 0,
                latencies: VecDeque::with_capacity(max_latency_samples),
                heartbeat_rtts: VecDeque::with_capacity(100),
                heartbeat_misses: 0,
            })),
        }
    }

    /// Start tracking a request.
    ///
    /// Returns start timestamp for later end_request() call.
    pub async fn start_request(&self, _request_id: &str, _function: &str, _namespace: &str) -> Instant {
        let mut inner = self.inner.write().await;
        inner.requests_total += 1;
        inner.queue_depth += 1;
        inner.queue_max_depth = inner.queue_max_depth.max(inner.queue_depth);
        Instant::now()
    }

    /// End tracking a request.
    ///
    /// Returns latency in milliseconds.
    pub async fn end_request(&self, start_time: Instant, _request_id: &str, success: bool, _error: Option<String>) -> f64 {
        let latency_ms = start_time.elapsed().as_secs_f64() * 1000.0;

        let mut inner = self.inner.write().await;
        inner.queue_depth = inner.queue_depth.saturating_sub(1);

        if success {
            inner.requests_success += 1;
        } else {
            inner.requests_failed += 1;
        }

        // Store latency sample (circular buffer behavior)
        inner.latencies.push_back(latency_ms);
        if inner.latencies.len() > inner.max_latency_samples {
            inner.latencies.pop_front();
        }

        latency_ms
    }

    /// Record a circuit breaker trip event.
    pub async fn record_circuit_breaker_trip(&self) {
        let mut inner = self.inner.write().await;
        inner.circuit_breaker_trips += 1;
        inner.circuit_breaker_state = "open".to_string();
    }

    /// Record circuit breaker reset (closed).
    pub async fn record_circuit_breaker_reset(&self) {
        let mut inner = self.inner.write().await;
        inner.circuit_breaker_state = "closed".to_string();
    }

    /// Record circuit breaker entering half-open state.
    pub async fn record_circuit_breaker_half_open(&self) {
        let mut inner = self.inner.write().await;
        inner.circuit_breaker_state = "half-open".to_string();
    }

    /// Record a heartbeat round-trip time.
    pub async fn record_heartbeat_rtt(&self, rtt_ms: f64) {
        let mut inner = self.inner.write().await;
        inner.heartbeat_rtts.push_back(rtt_ms);
        if inner.heartbeat_rtts.len() > 100 {
            inner.heartbeat_rtts.pop_front();
        }
    }

    /// Record a missed heartbeat (timeout).
    pub async fn record_heartbeat_miss(&self) {
        let mut inner = self.inner.write().await;
        inner.heartbeat_misses += 1;
    }

    /// Get a point-in-time snapshot of all metrics.
    pub async fn snapshot(&self) -> MetricsSnapshot {
        let inner = self.inner.read().await;

        // Calculate percentiles
        let mut latencies: Vec<f64> = inner.latencies.iter().cloned().collect();
        let (latency_avg, latency_p50, latency_p95, latency_p99, latency_min, latency_max) =
            if !latencies.is_empty() {
                latencies.sort_by(|a, b| a.partial_cmp(b).unwrap());
                let n = latencies.len();
                let p50_idx = (n as f64 * 0.50) as usize;
                let p95_idx = (n as f64 * 0.95) as usize;
                let p99_idx = (n as f64 * 0.99) as usize;

                let latency_avg = latencies.iter().sum::<f64>() / n as f64;
                let latency_p50 = latencies[p50_idx.min(n - 1)];
                let latency_p95 = latencies[p95_idx.min(n - 1)];
                let latency_p99 = latencies[p99_idx.min(n - 1)];
                let latency_min = latencies[0];
                let latency_max = latencies[n - 1];

                (latency_avg, latency_p50, latency_p95, latency_p99, latency_min, latency_max)
            } else {
                (0.0, 0.0, 0.0, 0.0, 0.0, 0.0)
            };

        // Calculate heartbeat RTT average
        let heartbeat_rtts: Vec<f64> = inner.heartbeat_rtts.iter().cloned().collect();
        let (heartbeat_rtt_avg, heartbeat_rtt_last) = if !heartbeat_rtts.is_empty() {
            let avg = heartbeat_rtts.iter().sum::<f64>() / heartbeat_rtts.len() as f64;
            let last = *heartbeat_rtts.last().unwrap();
            (avg, last)
        } else {
            (0.0, 0.0)
        };

        MetricsSnapshot {
            requests_total: inner.requests_total,
            requests_success: inner.requests_success,
            requests_failed: inner.requests_failed,
            latency_avg_ms: latency_avg,
            latency_p50_ms: latency_p50,
            latency_p95_ms: latency_p95,
            latency_p99_ms: latency_p99,
            latency_min_ms: latency_min,
            latency_max_ms: latency_max,
            queue_depth: inner.queue_depth,
            queue_max_depth: inner.queue_max_depth,
            circuit_breaker_trips: inner.circuit_breaker_trips,
            circuit_breaker_state: inner.circuit_breaker_state.clone(),
            heartbeat_rtt_avg_ms: heartbeat_rtt_avg,
            heartbeat_rtt_last_ms: heartbeat_rtt_last,
            heartbeat_misses: inner.heartbeat_misses,
            window_seconds: inner.window_seconds,
            timestamp: current_timestamp(),
        }
    }

    /// Reset all metrics.
    pub async fn reset(&self) {
        let mut inner = self.inner.write().await;
        inner.requests_total = 0;
        inner.requests_success = 0;
        inner.requests_failed = 0;
        inner.circuit_breaker_trips = 0;
        inner.circuit_breaker_state = "closed".to_string();
        inner.queue_depth = 0;
        inner.queue_max_depth = 0;
        inner.latencies.clear();
        inner.heartbeat_rtts.clear();
        inner.heartbeat_misses = 0;
    }
}

impl Default for Metrics {
    fn default() -> Self {
        Self::new()
    }
}

fn current_timestamp() -> f64 {
    use std::time::{SystemTime, UNIX_EPOCH};
    SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .map(|d| d.as_secs_f64())
        .unwrap_or(0.0)
}
